### ISSUE TITLE
Need help to understand what causes memory to be allocated

### ISSUE THREAD

**Issue by sergei-maertens** [unbit/uwsgi#2658]
tl;dr: in a docker container uwsgi is consuming 8GB (26%) of my system memory, which it doesn't outside the container and it only seems to happen on my system, not on those of my coworkers.

I'm running into excessive memory usage issues on my system when running uwsgi in docker (compose) containers and I can't figure out why this is happening and whether it's mis-configuration on my part or possibly a bug in uwsgi. I've been able to narrow down the problem a bit.

This *may* be related to #1818

**Environment/system**

* OS: Arch Linux
* Kernel: 6.10.0-arch1-2
* Docker version 27.0.3, build 7d4bcd863a
* Docker Compose version 2.29.1
* uwsgi 2.0.26
* 8 CPU cores, 16 threads (htop reports 16 cpus)
* 32GB of system memory (htop reports 30.6G)

**Problem**

Running uwsgi in a container consumes 8GB of memory. Memory usage is observed on the host system with `htop`. Running uwsgi on the host system itself consumes barely any memory (percentage in htop gets rounded to 0.0%):

```bash
uwsgi --http :8000
```

Interestingly, instead of `--http`, `--http-socket` and `--socket` both do not demonstrate this problem (VIRT memory usage sits around 18000), while `--http` uses 8GB of VIRT memory.

**Minimal reproduction**

```bash
docker run -it --rm debian:bookworm-slim bash
apt-get update && apt-get install build-essential python3 python3-pip
pip install uwsgi --break-system-packages  # debian uwsgi is 2.0.21 and modified
/usr/local/bin/uwsgi --http :8000
```

Output:

```
*** Starting uWSGI 2.0.26 (64bit) on [Thu Aug  1 10:06:36 2024] ***
compiled with version: 12.2.0 on 01 August 2024 10:04:32
os: Linux-6.10.0-arch1-2 #1 SMP PREEMPT_DYNAMIC Mon, 22 Jul 2024 17:28:23 +0000
nodename: 43f79679ec27
machine: x86_64
clock source: unix
detected number of CPU cores: 16
current working directory: /
detected binary path: /usr/local/bin/uwsgi
!!! no internal routing support, rebuild with pcre support !!!
uWSGI running as root, you can use --uid/--gid/--chroot options
*** WARNING: you are running uWSGI as root !!! (use the --uid flag) *** 
*** WARNING: you are running uWSGI without its master process manager ***
your memory page size is 4096 bytes
detected max file descriptor number: 1073741816
lock engine: pthread robust mutexes
thunder lock: disabled (you can enable it with --thunder-lock)
uWSGI http bound on :8000 fd 4
spawned uWSGI http 1 (pid: 5509)
uwsgi socket 0 bound to TCP address 127.0.0.1:35573 (port auto-assigned) fd 3
uWSGI running as root, you can use --uid/--gid/--chroot options
*** WARNING: you are running uWSGI as root !!! (use the --uid flag) *** 
Python version: 3.11.2 (main, May  2 2024, 11:59:08) [GCC 12.2.0]
*** Python threads support is disabled. You can enable it with --enable-threads ***
Python main interpreter initialized at 0x756ea6b60018
uWSGI running as root, you can use --uid/--gid/--chroot options
*** WARNING: you are running uWSGI as root !!! (use the --uid flag) *** 
your server socket listen backlog is limited to 100 connections
your mercy for graceful operations on workers is 60 seconds
mapped 72904 bytes (71 KB) for 1 cores
*** Operational MODE: single process ***
*** no app loaded. going in full dynamic mode ***
uWSGI running as root, you can use --uid/--gid/--chroot options
*** WARNING: you are running uWSGI as root !!! (use the --uid flag) *** 
*** uWSGI is running in multiple interpreter mode ***
spawned uWSGI worker 1 (and the only) (pid: 5508, cores: 1)
```

So, I'm not even loading any application and I'm observing this memory usage:

![image](https://github.com/user-attachments/assets/6280e930-bbae-47bc-8f2d-377ead0c9314)

Whether an app is loaded or not has the same outcome.

**What I've tried**

I've tried to `strace` uwsgi and checking for memory allocations. While I see a lot of them, I can't figure out where they're coming from - I don't have much familiarity with strace so I'm a bit lost here on what's happening.

**Comment by sergei-maertens** [comment]
Okay, one of my coworkers also runs Arch, so that gave a better comparison. He is *not* experiencing this problem:

```
*** Starting uWSGI 2.0.26 (64bit) on [Thu Aug  1 10:25:38 2024] ***
compiled with version: 12.2.0 on 01 August 2024 10:25:08
os: Linux-6.10.0-arch1-2 #1 SMP PREEMPT_DYNAMIC Mon, 22 Jul 2024 17:28:23 +0000
nodename: 5e8d09df5e5a
machine: x86_64
clock source: unix
detected number of CPU cores: 12
current working directory: /
detected binary path: /usr/local/bin/uwsgi
!!! no internal routing support, rebuild with pcre support !!!
uWSGI running as root, you can use --uid/--gid/--chroot options
*** WARNING: you are running uWSGI as root !!! (use the --uid flag) ***
*** WARNING: you are running uWSGI without its master process manager ***
your processes number limit is 60646
your memory page size is 4096 bytes
detected max file descriptor number: 524288
lock engine: pthread robust mutexes
thunder lock: disabled (you can enable it with --thunder-lock)
uWSGI http bound on :8000 fd 4
spawned uWSGI http 1 (pid: 5247)
uwsgi socket 0 bound to TCP address 127.0.0.1:33651 (port auto-assigned) fd 3
uWSGI running as root, you can use --uid/--gid/--chroot options
*** WARNING: you are running uWSGI as root !!! (use the --uid flag) ***
Python version: 3.11.2 (main, May  2 2024, 11:59:08) [GCC 12.2.0]
*** Python threads support is disabled. You can enable it with --enable-threads ***
Python main interpreter initialized at 0x79ae6cb60018
uWSGI running as root, you can use --uid/--gid/--chroot options
*** WARNING: you are running uWSGI as root !!! (use the --uid flag) ***
your server socket listen backlog is limited to 100 connections
your mercy for graceful operations on workers is 60 seconds
mapped 72904 bytes (71 KB) for 1 cores
*** Operational MODE: single process ***
*** no app loaded. going in full dynamic mode ***
uWSGI running as root, you can use --uid/--gid/--chroot options
*** WARNING: you are running uWSGI as root !!! (use the --uid flag) ***
*** uWSGI is running in multiple interpreter mode ***
spawned uWSGI worker 1 (and the only) (pid: 5246, cores: 1)
```

Things that stand out to me are:

* I don't have a process number limit, his is `your processes number limit is 60646`
* My max file descriptors is much higher, his is `detected max file descriptor number: 524288`

**Comment by sergei-maertens** [comment]
Okay, limiting the open file descriptors resolves the excessive memory usage... I suppose this makes sense in how the http router works for accepting incoming connections.

Closing this, my problem seems solved.
